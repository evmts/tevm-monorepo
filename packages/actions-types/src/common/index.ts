export type { Abi } from './Abi.js'
export type { Block } from './Block.js'
export type { BlockParam } from './BlockParam.js'
export type { BlockTag } from './BlockTag.js'
export type { Address } from './Address.js'
export type { Hex } from './Hex.js'
export type { EmptyParams } from './EmptyParams.js'
export type { Log } from './Log.js'
export type { BlockResult } from './BlockResult.js'
export type { FilterLog } from './FilterLog.js'
export type { FilterParams } from './FilterParams.js'
export type { TransactionParams } from './TransactionParams.js'
export type { TransactionResult } from './TransactionResult.js'
export type { TransactionReceiptResult } from './TransactionReceiptResult.js'
export type { TraceCall } from './TraceCall.js'
export type { TraceResult } from './TraceResult.js'
export type { TraceType } from './TraceType.js'
export type { NetworkConfig } from './NetworkConfig.js'
