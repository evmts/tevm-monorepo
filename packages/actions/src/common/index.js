export { isValidCallEvents, validateCallEvents } from './zCallEvents.js'
