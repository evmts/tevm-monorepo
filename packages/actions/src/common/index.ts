export type { Abi } from './Abi.js'
export type { AccountState } from './AccountState.js'
export type { Address } from './Address.js'
export type { Block } from './Block.js'
export type { BlockOverrideSet } from './BlockOverrideSet.js'
export type { BlockParam } from './BlockParam.js'
export type { BlockResult } from './BlockResult.js'
export type { BlockTag } from './BlockTag.js'
export type { CallEvents, Message, NewContractEvent } from './CallEvents.js'
export type { CallTraceResult } from './CallTraceResult.js'
export type { EmptyParams } from './EmptyParams.js'
export type { FlatCallTraceResult, FlatCallTraceCall } from './FlatCallTraceResult.js'
export type { FilterLog } from './FilterLog.js'
export type { FilterParams } from './FilterParams.js'
export type { Hex } from './Hex.js'
export type { Log } from './Log.js'
export type { NetworkConfig } from './NetworkConfig.js'
export type { PrestateTraceResult } from './PrestateTraceResult.js'
export type { StateOverrideSet } from './StateOverrideSet.js'
export type { StructLog } from './StructLog.js'
export type { TraceCall } from './TraceCall.js'
export type { TraceResult } from './TraceResult.js'
export type { TraceType } from './TraceType.js'
export type { TransactionParams } from './TransactionParams.js'
export type { TransactionReceiptResult } from './TransactionReceiptResult.js'
export type { TransactionResult } from './TransactionResult.js'
