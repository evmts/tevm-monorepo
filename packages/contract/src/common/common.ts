export { ERC2771Context, ERC2771Forwarder } from './metatx'
export { Multicall, Multicall3 } from './multicall'
export {
	ERC1155,
	ERC20,
	ERC20Votes,
	ERC2981,
	ERC721,
	ERC721Votes,
} from './tokens'
