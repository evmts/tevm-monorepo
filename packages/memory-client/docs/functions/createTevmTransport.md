[**@tevm/memory-client**](../README.md) • **Docs**

***

[@tevm/memory-client](../globals.md) / createTevmTransport

# Function: createTevmTransport()

> **createTevmTransport**(`options`): [`TevmTransport`](../type-aliases/TevmTransport.md)\<`string`\>

## Parameters

• **options**: `TevmNodeOptions`\<`Common`\> = `{}`

Configuration options for the base client, similar to those used in `memoryClient` or a low-level `baseClient`.

## Returns

[`TevmTransport`](../type-aliases/TevmTransport.md)\<`string`\>

A configured TEVM transport.

## Defined in

[packages/memory-client/src/createTevmTransport.js:58](https://github.com/evmts/tevm-monorepo/blob/main/packages/memory-client/src/createTevmTransport.js#L58)
