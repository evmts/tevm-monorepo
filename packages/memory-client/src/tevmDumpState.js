import { dumpStateHandler } from '@tevm/actions'

/**
 * A tree-shakeable version of the `tevmDumpState` action for viem.
 * Dumps the state of TEVM into a plain JavaScript object that can later be used to restore state
 * using the `tevmLoadState` action.
 *
 * This action is useful for persisting and restoring the state between different sessions or processes.
 *
 * @param {import('viem').Client<import('./TevmTransport.js').TevmTransport<string>>} client - The viem client configured with TEVM transport.
 * @returns {Promise<import('@tevm/actions').DumpStateResult>} The dump of the TEVM state.
 *
 * @example
 * ```typescript
 * import { tevmDumpState } from 'tevm/actions'
 * import { createClient, http } from 'viem'
 * import { optimism } from 'tevm/common'
 * import { createTevmTransport } from 'tevm'
 * import fs from 'fs'
 *
 * const client = createClient({
 *   transport: createTevmTransport({
 *     fork: { transport: http('https://mainnet.optimism.io')({}) }
 *   }),
 *   chain: optimism,
 * })
 *
 * async function example() {
 *   const state = await tevmDumpState(client)
 *   fs.writeFileSync('state.json', JSON.stringify(state))
 *   console.log('State dumped to state.json')
 * }
 *
 * example()
 * ```
 *
 * @see [DumpStateResult](https://tevm.sh/reference/tevm/actions/type-aliases/dumpstateresult/) for return values reference.
 * @see [TEVM Actions Guide](https://tevm.sh/learn/actions/)
 * @see [tevmLoadState](https://tevm.sh/reference/tevm/actions/functions/tevmLoadState/) for restoring the state.
 */
export const tevmDumpState = async (client) => {
	return dumpStateHandler(client.transport.tevm)()
}
