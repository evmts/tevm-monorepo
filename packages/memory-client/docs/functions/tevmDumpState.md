[**@tevm/memory-client**](../README.md)

***

[@tevm/memory-client](../globals.md) / tevmDumpState

# Function: tevmDumpState()

> **tevmDumpState**(`client`): `Promise`\<`DumpStateResult`\<`TevmDumpStateError`\>\>

Defined in: [packages/memory-client/src/tevmDumpState.js:120](https://github.com/evmts/tevm-monorepo/blob/main/packages/memory-client/src/tevmDumpState.js#L120)

A tree-shakeable version of the `tevmDumpState` action for viem.
Exports the entire blockchain state from TEVM into a serializable JavaScript object.

This function captures a complete snapshot of the current blockchain state, including:
- All account balances, nonces, and storage
- Smart contract bytecode
- Transaction receipts
- Block headers and history
- Current blockchain parameters (latest block, gas price, etc.)

The returned state object can be used for:
- Persisting state between application sessions
- Backing up blockchain state before running test scenarios
- Transferring state between different TEVM instances
- Creating reproducible environments for debugging
- Defining pre-configured test fixtures

This function works in conjunction with `tevmLoadState` to provide a complete state
serialization and restoration system. The output is serializable using standard
JSON.stringify() though you may need to handle BigInt conversion depending on your
platform (see the example below).

## Parameters

### client

`Client`\<[`TevmTransport`](../type-aliases/TevmTransport.md)\<`string`\>, `undefined` \| `Chain`, `undefined` \| `Account`, `undefined`, `undefined` \| \{[`key`: `string`]: `unknown`; `account?`: `undefined`; `batch?`: `undefined`; `cacheTime?`: `undefined`; `ccipRead?`: `undefined`; `chain?`: `undefined`; `key?`: `undefined`; `name?`: `undefined`; `pollingInterval?`: `undefined`; `request?`: `undefined`; `transport?`: `undefined`; `type?`: `undefined`; `uid?`: `undefined`; \}\>

The viem client configured with TEVM transport.

## Returns

`Promise`\<`DumpStateResult`\<`TevmDumpStateError`\>\>

A serializable object containing the complete blockchain state.

## Examples

```typescript
import { tevmDumpState, tevmLoadState } from 'tevm/actions'
import { createClient, http, parseEther } from 'viem'
import { optimism } from 'tevm/common'
import { createTevmTransport } from 'tevm'
import fs from 'fs/promises'

const client = createClient({
  transport: createTevmTransport({
    fork: { transport: http('https://mainnet.optimism.io')({}) }
  }),
  chain: optimism,
})

async function saveAndRestoreState() {
  // Set up initial state
  await client.setBalance({
    address: '0x1234567890123456789012345678901234567890',
    value: parseEther('100')
  })
  await client.mine() // Mine a block to update state

  // Dump the entire blockchain state
  const state = await tevmDumpState(client)

  // Serialize state to JSON (handling BigInt conversion)
  const serializedState = JSON.stringify(state, (_, value) =>
    typeof value === 'bigint' ? value.toString() : value
  )

  // Save to file
  await fs.writeFile('blockchain-snapshot.json', serializedState)
  console.log('State saved to blockchain-snapshot.json')

  // Later, in another session or process:

  // Read the state from file
  const savedState = JSON.parse(
    await fs.readFile('blockchain-snapshot.json', 'utf8'),
    (_, value) => {
      // Restore BigInt values
      if (typeof value === 'string' && /^\d+n$/.test(value)) {
        return BigInt(value.slice(0, -1))
      }
      return value
    }
  )

  // Create a new client with the saved state
  const newClient = createClient({
    transport: createTevmTransport(),
    chain: optimism,
  })

  // Load the saved state
  await tevmLoadState(newClient, savedState)

  // Verify the state was restored
  const balance = await newClient.getBalance({
    address: '0x1234567890123456789012345678901234567890'
  })
  console.log('Restored balance:', balance) // Should be 100 ETH
}
```

```typescript
// Using with the SyncStoragePersister for automatic persistence
import { createMemoryClient, createSyncPersister } from 'tevm'

// Create a persister that saves to localStorage (in browser)
// or to a file (in Node.js)
const persister = createSyncPersister({
  storage: localStorage, // or use a file-based storage adapter
  key: 'tevm-state'      // storage key
})

// Create a client with automatic persistence
const client = createMemoryClient({
  persister,
  // State will be automatically loaded from storage on startup
  // and saved to storage periodically and on shutdown
})
```

## See

 - [DumpStateResult](https://tevm.sh/reference/tevm/actions/type-aliases/dumpstateresult/) for return value structure.
 - [TEVM Actions Guide](https://tevm.sh/learn/actions/)
 - [tevmLoadState](https://tevm.sh/reference/tevm/actions/functions/tevmloadstate/) for restoring the state.
 - [SyncStoragePersister](https://tevm.sh/reference/tevm/sync-storage-persister/functions/createsyncpersister/) for automatic state persistence.
