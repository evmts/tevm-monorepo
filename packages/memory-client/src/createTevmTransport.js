import { createCommon } from '@tevm/common'
import { requestEip1193, tevmSend } from '@tevm/decorators'
import { createTevmNode } from '@tevm/node'
import { createTransport } from 'viem'

/**
 * Creates a custom TEVM Transport for viem.
 *
 * A Transport in viem is the intermediary layer responsible for executing outgoing RPC requests. This custom TEVM Transport integrates an in-memory Ethereum client, making it ideal for local-first applications, optimistic updates, and advanced TEVM functionalities like scripting.
 *
 * @param {import('@tevm/node').TevmNodeOptions} options - Configuration options for the base client, similar to those used in `memoryClient` or a low-level `baseClient`.
 * @returns {import('./TevmTransport.js').TevmTransport} A configured TEVM transport.
 *
 * @example
 * ```typescript
 * import { createClient, http } from 'viem'
 * import { createTevmTransport } from 'tevm'
 * import { optimism } from 'tevm/common'
 *
 * const client = createClient({
 *   transport: createTevmTransport({
 *     fork: { transport: http('https://mainnet.optimism.io')({}) }
 *   }),
 *   chain: optimism,
 * })
 *
 * async function example() {
 *   const blockNumber = await client.getBlockNumber()
 *   console.log(blockNumber)
 * }
 *
 * example()
 * ```
 *
 * @see {@link createClient}
 * @see [Viem Client Docs](https://viem.sh/docs/clients/introduction)
 * @see [Client Guide](https://tevm.sh/learn/clients/)
 * @see [tevm JSON-RPC Guide](https://tevm.sh/learn/json-rpc/)
 * @see [EIP-1193 spec](https://eips.ethereum.org/EIPS/eip-1193)
 * @see [Ethereum jsonrpc docs](https://ethereum.org/en/developers/docs/apis/json-rpc/)
 * @see [CreateMemoryClient Docs](https://tevm.sh/reference/tevm/memory-client/functions/creatememoryclient/) - For a batteries-included client if not worried about tree shaking
 *
 * @typedef {Object} TevmTransportConfig
 * @property {number} [timeout=20000] - Timeout duration for requests in milliseconds. Default is 20,000 ms. Supplied by viem.
 * @property {number} [retryCount=3] - The maximum number of times to retry a failed request. Default is 3. Supplied by viem.
 * @property {import('viem').Chain} [chain] - Blockchain configuration. Defaults to the chain specified in `options` or the default TEVM chain.
 *
 * ## Parameters
 *
 * - `timeout` (optional, number): Timeout duration for requests in milliseconds. Default is 20,000 ms. Supplied by viem.
 * - `retryCount` (optional, number): The maximum number of times to retry a failed request. Default is 3. Supplied by viem.
 * - `chain` (optional, Chain): Blockchain configuration. Defaults to the chain specified in `options` or the default TEVM chain if not provided.
 *
 * ## Gotchas
 *
 * - When specifying a chain, use TEVM common instead of viem chains. You can create a TEVM common from a viem chain using `createCommon`.
 */
export const createTevmTransport = (options = {}) => {
	/**
	 * A map to store and manage TEVM clients keyed by chain ID.
	 * @type {Map<number, import('@tevm/node').TevmNode & import('@tevm/decorators').Eip1193RequestProvider & import('@tevm/decorators').TevmSendApi>}
	 */
	const tevmMap = new Map()

	/**
	 * Creates and returns a TEVM transport.
	 * @type {import('./TevmTransport.js').TevmTransport}
	 */
	return ({ timeout = 20_000, retryCount = 3, chain }) => {
		const dynamicChain =
			chain && 'ethjsCommon' in chain
				? /** @type {import('@tevm/common').Common} */ (chain)
				: chain !== undefined
					? // if user passed in chain instead of common create a common from it with cancun and default eips
						createCommon({ ...chain, hardfork: 'cancun', loggingLevel: 'warn' })
					: undefined
		const common = options.common ?? dynamicChain
		const id = common?.id ?? -1
		const tevm =
			tevmMap.get(id) ??
			createTevmNode({ ...options, ...(common !== undefined ? { common } : {}) })
				.extend(requestEip1193())
				.extend(tevmSend())
		tevmMap.set(id, tevm)

		return /** @type {any} */ (
			createTransport(
				{
					request: /** @type any */ (tevm.request),
					type: 'tevm',
					name: /** options?.name ?? */ 'Tevm transport',
					key: /* options?.key ?? */ 'tevm',
					timeout,
					retryCount,
					retryDelay: /* options?.retryDelay ?? */ 150,
				},
				{ tevm },
			)
		)
	}
}
