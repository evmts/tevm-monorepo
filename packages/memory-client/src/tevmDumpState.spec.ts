import { optimism } from '@tevm/common'
import { transports } from '@tevm/test-utils'
import { type Client, createClient } from 'viem'
import { beforeEach, describe, expect, it } from 'vitest'
import type { TevmTransport } from './TevmTransport.js'
import { createTevmTransport } from './createTevmTransport.js'
import { tevmDumpState } from './tevmDumpState.js'

let client: Client<TevmTransport>

beforeEach(async () => {
	client = createClient({
		transport: createTevmTransport({
			fork: { transport: transports.optimism },
		}),
		chain: optimism,
	})
})

describe('tevmDumpState', () => {
	it('should dump the state of the TEVM', async () => {
		const stateDump = await tevmDumpState(client)
		expect(stateDump).toMatchSnapshot()
	})
})
