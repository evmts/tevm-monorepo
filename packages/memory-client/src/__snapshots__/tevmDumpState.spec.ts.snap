// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tevmDumpState > should dump the state of the TEVM 1`] = `
{
  "state": {
    "0x0000000000000000000000000000000000000000": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x14dC79964da2C08b23698B3D3cc7Ca32193d9955": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x23618e81E3f5cdF7f54C3d65f7FBc0aBf5B21E8f": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x70997970C51812dc3A010C7d01b50e0d17dc79C8": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x90F79bf6EB2c4f870365E785982E1f101E93b906": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x976EA74026E726554dB657fA54763abd0C3a0aa9": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0x9965507D1a55bcC2695C58ba16FB37d819B0A4dc": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0xa0Ee7A142d267C1f36714E4a8F75612F20a79720": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266": {
      "balance": 1000000000000000000000n,
      "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce": 0n,
      "storage": {},
      "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
  },
}
`;
